### PR TITLE
Add bulk inserting/updating list of students

### PR DESCRIPTION
**Pull request by slavo-furman**
This PR contains two commits. 

First is updating the userservice app by adding a new http api for bulk inserting/updating list of students. This new http api is implemented by extending /user/ endpoint and allow call it via http POST where list of student is payload in body of http message. In case of success it returns 201 Created status code (I know it is kind of controversial topic, but I prefer that all http api calls return clear status code). Underlying service method simply use C# Azure SDK for upserting Azure Table Storage.

Second adds web ui (simple bootstrap styled html input form) to /admin/users page of presentation app. User is expected to insert list of email addresses, one email address per line, no explicit separators, along with claim name and claim value which will be applied in same way to each email address. Email addresses are validated before we pass them to underlying service. Email validator is implemented in EmailValidator class. Underlying service method, implemented in AuthenticationService class simply forms json payload from passed list of students and calls http api endpoint implemented in userservice app.

### CODE REVIEW

**Review by koscelansky — CHANGES_REQUESTED** [review]

File: cppseminar/presentation/Model/EmailValidator.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                return false;
+            }
+            catch (ArgumentException e)
+            {
+                return false;
+            }
+
+            try
+            {
+                return Regex.IsMatch(email,
```
[the remark this answers is not in the document]
Comment (slavo-furman) [reply cppseminar/presentation/Model/EmailValidator.cs:48]: I agree. Changed.

File: cppseminar/userservice/Controllers/UserController.cs [hunk 2]
```diff
@@ -35,5 +35,21 @@ public async Task<ActionResult<UserModel>> GetUserClaims([FromRoute]string email
         {
             return await _service.GetUserAync(email);
         }
+
+        [HttpPost]
+        public async Task<IActionResult> UpdateListOfStudents([FromBody] List<UserModel> listOfStudents)
```
[the remark this answers is not in the document]
Comment (slavo-furman) [reply cppseminar/userservice/Controllers/UserController.cs:40]: All mentions of "students" in identifiers were changed to "users".

File: cppseminar/userservice/Services/UserDbService.cs [hunk 3]
```diff
@@ -64,9 +64,27 @@ public async Task<UserModel> GetUserAync(string userEmail)
             };
         }
 
-        public async Task CreateUserAsync(UserModel user)
+        public async Task UpdateListOfStudents(List<UserModel> listOfStudents)
         {
+            try
+            {
+                for (int i = 0; i < listOfStudents.Count; i++)
```
[the remark this answers is not in the document]
Comment (slavo-furman) [reply cppseminar/userservice/Services/UserDbService.cs:71]: Changed to foreach loop.

File: cppseminar/presentation/Pages/Shared/_Layout.cshtml [hunk 4]
```diff
@@ -49,6 +49,8 @@
         @RenderBody()
     </div>
     <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-validate/1.19.4/jquery.validate.min.js" integrity="sha512-FOhq9HThdn7ltbK8abmGn60A/EMtEzIzv1rvuh+DqzJtSGq8BRdEN0U+j0iKEIffiw/yEtVuladk6rsG4X6Uqg==" crossorigin="anonymous"></script>
```
[the remark this answers is not in the document]
Comment (slavo-furman) [reply cppseminar/presentation/Pages/Shared/_Layout.cshtml:52]: I originally thought that this is needed for form validation to work. But it seems that at least basic validation is working fine without it. Removed.

File: cppseminar/presentation/Pages/Admin/Users/Index.cshtml.cs [hunk 5]
```diff
@@ -17,14 +17,25 @@ public class IndexModel : PageModel
     {
         private ILogger<IndexModel> _logger;
         private AuthenticationService _authService;
+
         public IEnumerable<string> AllUsers = Enumerable.Empty<string>();
         public IDictionary<string, string> Claims { get; set; }
 
+        [BindProperty]
+        public string studentlist { get; set; }
```
[the remark this answers is not in the document]
Comment (slavo-furman) [reply cppseminar/presentation/Pages/Admin/Users/Index.cshtml.cs]: Model, service and input form for update list of users were completely rewritten as requested, so given question is no longer relevant.

File: cppseminar/presentation/Pages/Admin/Users/Index.cshtml.cs [hunk 6]
```diff
@@ -61,5 +72,67 @@ public async Task OnGetDetail([FromQuery][Required]string email)
...
         }
 
+        public async Task<IActionResult> OnPost()
+        {
+            foreach (KeyValuePair<string, string> entry in
+                     new Dictionary<string, string> { { "claimname", "Claim name" },
+                                                      { "claimvalue", "Claim value" },
+                                                      { "studentlist", "Student list" }})
+            {
+                if (string.IsNullOrWhiteSpace(Request.Form[entry.Key].ToString()))
```
[the remark this answers is not in the document]
Comment (slavo-furman) [reply cppseminar/presentation/Pages/Admin/Users/Index.cshtml.cs]: Model, service and input form for update list of users were completely rewritten as requested, so given question is no longer relevant.

File: cppseminar/presentation/Pages/Admin/Users/Index.cshtml [hunk 7]
```diff
@@ -1,16 +1,66 @@
...
+      <div class="col-sm-5">          
+        <input type="text" class="form-control" id="claimvalue" placeholder="Enter claim value" name="claimvalue"  value="@Model.claimvalue">
+      </div>
+    </div>
+    <div class="form-group">        
+      <div class="col-sm-offset-2 col-sm-10">
+        <button type="submit" class="btn btn-primary">Update student list</button>
+      </div>
+    </div>
+    @Html.AntiForgeryToken()
```
[the remark this answers is not in the document]
Comment (slavo-furman) [reply cppseminar/presentation/Pages/Admin/Users/Index.cshtml]: Model, service and input form for update list of users were completely rewritten as requested, so given question is no longer relevant.

File: cppseminar/userservice/Controllers/UserController.cs [hunk 8]
```diff
@@ -35,5 +35,21 @@ public async Task<ActionResult<UserModel>> GetUserClaims([FromRoute]string email
...
+        {
+            try
+            {
+                await _service.UpdateListOfUsers(listOfUsers);
+
+                return StatusCode(201);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"Problem in [UploadListOfUsers] action: {ex.Message}");
```
Comment (lukas-manduch) [inline cppseminar/userservice/Controllers/UserController.cs:50]: Preco je tu pouzity iny string builder pre logger nez ten standradny? Myslim $ na zaciatku. Bud by som pouzival jeden sposob, alebo druhy

File: cppseminar/userservice/Controllers/UserController.cs [hunk 9]
```diff
@@ -35,5 +35,21 @@ public async Task<ActionResult<UserModel>> GetUserClaims([FromRoute]string email
...
+        [HttpPost]
+        public async Task<IActionResult> UpdateListOfUsers([FromBody] List<UserModel> listOfUsers)
+        {
+            try
+            {
+                await _service.UpdateListOfUsers(listOfUsers);
+
+                return StatusCode(201);
+            }
+            catch (Exception ex)
```
Comment (lukas-manduch) [inline cppseminar/userservice/Controllers/UserController.cs:48]: Tento catch podla mna nemoze fungovat, lebo vo funkcii co volas je tiez globalny catch

File: cppseminar/presentation/Services/UserService.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            _client.DefaultRequestHeaders.Accept.Clear();
+            _client.DefaultRequestHeaders.Accept.Add(
+                new MediaTypeWithQualityHeaderValue("application/json"));
+        }
+
+        public async Task OnUpdateUserListAsync(List<UserRest> userList)
+        {
+            string userListAsJson = JsonSerializer.Serialize(userList);
+
+            var stringContent = new StringContent(userListAsJson, Encoding.UTF8, "application/json");
```
Comment (lukas-manduch) [inline cppseminar/presentation/Services/UserService.cs:32]: Aky ma zmysel definovat dva krat to "application/json"? Myslim, preco je to aj tu aj o 5 riadkov vyssie

File: cppseminar/presentation/Services/AuthenticationService.cs [hunk 11]
```diff
@@ -7,6 +7,8 @@
 using Microsoft.AspNetCore.Authentication.OAuth;
 using Microsoft.Extensions.Configuration;
 using presentation.Model;
+using System.Text;
```
Comment (lukas-manduch) [inline cppseminar/presentation/Services/AuthenticationService.cs:10]: Neviem ci mi to github zobrazuje spravne, ale toto vyzera ako keby si pridal nejake veci navyse do suboru, s ktorym si nic nerobil

File: cppseminar/presentation/Pages/Admin/Users/Create.cshtml.cs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+using Microsoft.Extensions.Logging;
+using presentation.Model;
+using presentation.Services;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace presentation.Pages.Admin.Users
+{
+    [Authorize(Policy = "Administrator")]
```
Comment (lukas-manduch) [inline cppseminar/presentation/Pages/Admin/Users/Create.cshtml.cs:13]: Toto tu urcite nema byt. V ziadnom inom subore to tiez nie je pouzite

File: cppseminar/presentation/Model/EmailValidator.cs [hunk 13]
```diff
@@ -0,0 +1,51 @@
+using System;
+using System.Globalization;
+using System.Text.RegularExpressions;
+
+namespace presentation.Model
```
Comment (lukas-manduch) [inline cppseminar/presentation/Model/EmailValidator.cs:5]: Nemyslim si ze validator by mal byt medzi Models . Tam su veci reprezentujuce data a vacsinou bez nejakej aplikacnej logiky. Toto je presny opak

File: cppseminar/presentation/Model/EmailValidator.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+                    string domainName = idn.GetAscii(match.Groups[2].Value);
+
+                    return match.Groups[1].Value + domainName;
+                }
+
+                return Regex.IsMatch(email,
+                                     @"^[^@\s]+@[^@\s]+\.[^@\s]+$",
+                                     RegexOptions.IgnoreCase, TimeSpan.FromMilliseconds(250));
+            }
+            catch (RegexMatchTimeoutException)
```
Comment (lukas-manduch) [inline cppseminar/presentation/Model/EmailValidator.cs:41]: Toto je trocha zbytocny catch navyse, nie?

File: cppseminar/presentation/Pages/Admin/Users/Index.cshtml [hunk 15]
```diff
@@ -1,16 +1,28 @@
...
+        @foreach (var user in Model.AllUsers)
+        {
+            <tr>
+                <td><th:a asp-page="/Admin/Users/Index" asp-page-handler="Detail" asp-route-email="@user"> @user</th:a></td>            
+            </tr>
+        }
+    </table>
+
+    <div align="right">
+        <p><th:a asp-page="/Admin/Users/Create">Update user list</th:a></p>
```
Comment (lukas-manduch) [inline cppseminar/presentation/Pages/Admin/Users/Index.cshtml:23]: Trocha mi nie je jasne, preco vsade v texte je to update user list, zatial co v suboroch je to create. Pride mi to ako zbytocna nekonzistencia uz zo zaciatku

**Review by lukas-manduch — CHANGES_REQUESTED** [review]

File: cppseminar/userservice/Controllers/UserController.cs (hunk 9 above)
Comment (koscelansky) [reply cppseminar/userservice/Controllers/UserController.cs:48]: Tak toto nesuhlasim, ze je problem tu, ked tam nema byt v tej funkcii co sa vola.

**Pull request merged**